Add a MRAC preview post (#58)
ComplianceAsCode/ComplianceAsCode.github.io: _posts/2023-04-21-mrac.md

---
layout: post
title: "Change Management in the Content Project"
categories: template
author: Matěj Týč
author_url: https://github.com/matejak
---

Systems that our content aims to harden are very often large collections of separate components whose life cycle is independent from the life cycle of the product that ships them.

> For example, the Red Hat Enterprise Linux consists of more than six thousand of such distinct components, although only a fraction of them s.a. OpenSSH, Grub2 and so on is supported by the ComplianceAsCode project.

Reviewer: either such as or and so on but never both

We don't want the project's code polluted by difficult-to-read if-else conditions and copy-pasted snippets that undermine maintainability.

Anyway, when we talk about component changes, what do we have in mind?
Here are some examples:

- Configuration preference changes - instead of putting everything into one file, prefer distributing the configuration into a directory, for example prefer `sshd_config.d` directory over the `sshd_config` file.
- A configuration option is renamed, e.g. from a `whitelist` to `allowlist`.
- A feature disappears, and an alternative approach is needed - consider the case of OpenSSH losing capabilities of the `SetIdleTimeout 0` option to drop inactive sessions.

## New Approach Needed

If we think in a very simple way, we clearly need to design smart procedures or workflows that react to changes, so no major brain power is needed to make the right decisions.
However, in order to be able to come with workflows that introduce an added value, the project needs to gain some extra capabilities.

Our team has come up with these ideas so far:

- Lower the pressure on rules - enable coverage of a security requirement by more than one rule.

> - Be able to split a rule into such a set of rules easily.

Reviewer: ```suggestion
- Be able to split a rule into a set of rules easily.
```

- Enhance the declarative aspect of the project.
- Introduce a component-centric view to the project.
- Facilitate extension of a rule's scope.

## Closer Look

Let's take a look how those ideas could align to the subject of component changes:

### Multiple rules to address one requirement

> In practice, we may not know at build-time what exact version of a component will be scanned, and we may address this uncertainty by having multiple rules prepared to handle this uncertainty.

Reviewer: repeating uncertainity

### Go Declarative

Some areas of the project are already declarative, but to have more is better.
Declarative way of doing things allows for the introduction of more levels of abstraction, and when something changes, it is much more likely that a change will stay contained in one of such abstraction levels.

For example, we already have an enhanced declarative part in the form of [control files](https://github.com/ComplianceAsCode/content/tree/master/controls).
Thanks to that, we don't have to specify relations of rules to the profile directly, but we can focus on a much easier to grasp relation of a rule to a security control.
Profiles then can be [defined](https://github.com/ComplianceAsCode/content/blob/master/products/rhel8/profiles/cis_workstation_l2.profile) using security controls that stay, even though their rule composition may change.
As an additional benefit, the assignment of rules to security controls can be reused by other parts of the project.

However, this declarative concept can be extended beyond profile compositions.

> We often use constructs s.a. `{{% if product in ("rhel8", "rhel9") %}}` instead of a more generic and understandable `{{% if product.prefers_single_file_sshd_configuration %}}` or something analogous.

Reviewer: ```suggestion
We often use constructs s.a. `{{% if product in ("rhel8", "rhel9") %}}` instead of a more generic and understandable `{{% if product.prefers_single_file_sshd_configuration %}}` or similar.
```

In other words, we shouldn't miss an opportunity to declare that a certain product has a particular property, and every other part of the content should refer to those properties rather than to product names.
Designing such product properties in a way that is smart, doesn't get in the way and that can be reused in prose, checks and remediations is not trivial.

Can you think of other ways of bringing declarative principles to the project?

> ###  Be able to track component changes and our reactions to them.

Reviewer: remove the dot

Change "Be able" to "Ability"

However, later you discover that something is still not right, and you need to check out the reaction to that change.
What do you do?

You probably check the history of files that had to be touched at that time, and you are able to recover the exact date range of the change, and then related pull requests along with the discussion.
However, this information needs to be manually recovered from other changes to the project.
That's doable, but it is, to some degree, a detective work.

Wouldn't it be nice to have a capability that would allow us to query a component, and we would get change information from the project using some automation?
This can be very difficult to achieve exhaustively, but a partial and good enough solution may not be so difficult.

###  Polymorphic rules

Rules that try to do two things don't sound like a good idea, but if those two things are almost the same, then perhaps it could make sense.
Although `CCE` identifiers restrict the scope of rule polymorphism significantly, change of component configuration preference could be handled without splitting a rule into two, polymorphically.

> We have some polymorphic rules in our content, but they became polymorphic on their own, the project hasn't helped them with their polymorphism.

Reviewer: little cumbersome

For instance, the rule [no_empty_passwords](https://github.com/ComplianceAsCode/content/tree/master/linux_os/guide/system/accounts/accounts-restrictions/password_storage/no_empty_passwords) mentioned earlier has polymorphic Bash remediation and a rather simple OVAL check, while [enable_fips_mode](https://github.com/ComplianceAsCode/content/tree/master/linux_os/guide/system/software/integrity/fips/enable_fips_mode) has a simple Bash remediation, but a complex OVAL.

Provided examples nicely illustrate the richness of the rule-level polymorphism in the project.
Rules don't advertise that they are polymorphic, and the project doesn't provide them with any specific help with their polymorphism.

> Project support for polymorphism of rules could reduce maintenance costs, as polymorphism would become more declarative.

Reviewer: ```suggestion
Adding support for polymorphism of rules could reduce maintenance costs, as polymorphism would become more declarative.
```

---
layout: post
title: "Change Management in the Content Project"
categories: template
author: Matěj Týč
author_url: https://github.com/matejak
---

> Our Red Hat group is contemplating the idea of making the content project better equipped to handle changes in component behavior better.

Reviewer: s/component/software/

Author: Software is a too generic term, and I find component fitting.
I have added a description of what a component is, did it help?

As the project is open-source and has a community, we would like to get it involved from the very start.

## Introduction

Software changes all the time, not only when major version numbers change.
Sometimes it happens that what some consider a feature, developers see as a loophole, and when they close it, the component changes its behavior significantly.
Our project that configures components needs to accomodate for this change, and when it is done impulsively, without considering long-term impact on maintainability, those changes can result in difficult-to-read conditions, and in lots of copy-pasted code lying around.

Anyway, when we talk about changes, what do we have in mind?
Check out some examples:

- Configuration preference changes - instead of putting everything into one file, prefer partitioning the configuration into an e.g. `sshd_config.d` directory.
- A configuration option is renamed, e.g. from a whitelist to allowlist.
- A feature disappears, and alternative approach is needed - consider the case of OpenSSH losing capabilities of the `SetIdleTimeout 0` option. 

## Approach

> If we think in a very simple way, we clearly need to design smart procedures or workflows that recact to changes, so no major brain power is needed to make right decisions.

Reviewer: recact

However, in order to be able to come with workflows that introduce an added value, the project needs to gain some extra capabilities.

Our team has come up with these ideas so far:

- Be able to cover a security requirement by multiple rules.

> In practice, if we don't know at build-time what component will be scanned, we may solve this by having multiple rules ready.

Reviewer: not ready but having multiple rules where each rule is for a different version of the examined software

Author: reworded (differently), hopefully good enough.

When it comes to the actual scanning, this collection of rules makes sure that the evaluation will be carried out correctly, regardless of what component version is in question.

> This could be achieved by set of rules with disjoint applicability, and the recently-introduced CPE Applicability Language helps this.

Reviewer: demonstrate this on an example

- Be able to split a rule into such set of rules easily.

One could split rules by copy-pasting, but it is desirable to minimize the amount of manual operations to the minimum.
At the time of writing this post, creating a new rule involves noticeable boilerplate, and creating an alternative version of a rule requires more effort that necessary.
This has to be reversed - what can be automated should be automated, and the duplication of data has to decrease by applying a more aggressive normal form.
Specifically, references are defined in rule files, while they can also be deduced from controls.

> And perhaps there are other examples of redundant data?

Reviewer: Confusing. What are you trying to describe? I would recommend using templates and Jinja macros.

Demonstrate this on an example so that the readers can imagine it.

- Emphasize the declarative part of the project that will stay the same when a component changes.

We already have a declarative part in the form of control files.
When defining a profile using security controls, controls are the same even though their composition and underlying rules may change.

However, this declarative concept can be extended beyond profile compositions.
We often use constructs s.a. `{{% if product in ("rhel7", "rhel8") %}}` instead of a more generic `{{% if product.prefers_single_file_sshd_configuration %}}` or something analogous.

> Designing product properties in a way that is smart, doesn't get in the way and that can be reused in prose, checks and remediations is not trivial.

Reviewer: Please reword it that you recommend using product properties. Please add a working example.

- Be able to track component changes and our reactions to them.

Imagine that a component changes behavior, and you get pull requests that react to that change merged successfuly.
However, later you discover that something is still not right, and you need to check out that reaction.
What do you do?
You probably check the history of files that had to be touched at that time, and you are able to recover the date, and then related pull requests along with the discussion.
However, this information needs to be manually extracted from other changes to the project.
Wouldn't it be nice to have a capability that would allow us to query a component, and we would get change information from the project using an automation?

> This can be very difficult to achieve exhaustively, but partial/good enough achievement may be actually doable.

Reviewer: Is this a proposal for a future feature? If yes, introduce it as such and be more specific.

Author: A proposal will be introduced properly in the PR that implements it. This post just aims to shift attention to some gaps in the project, and to potential solutions that nobody knows right now.

- Be able to widen a scope of a rule, i.e. make it polymorphic.

Rules that try to do two things don't sound like a good idea, but if those two things are almost the same, then perhaps it could make sense.
Although CCE identifiers restrict the rule polymorphism significantly, change of component configuration preference could be handled without splitting a rule into two.

> 

Reviewer: example missing

## What now

Do you have something to add?
Or perhaps you would like to object some of those points?
Then reach out to us on Gitter!

> If things go ahead as planned, you may expect pull requests showing up in upcoming months, and we are very much interested in your point of view.

Reviewer: What will these PRs introduce?

Author: Nobody knows now, so we have to wait for them with great anticipation.
